refactor(mobile): local albums

shenlong-tanwen opened the pull request:
#### Changes made

- Refactored the mobile app handling to have separate collections for Local Albums and Remote albums
- Currently, all the local albums and assets are fetched and hashed and displayed in the main timeline. This will be addressed in a successive PR to select which assets to list in the main timeline (All Local / Selected / None)
- Separate Album Viewer Widgets for Local and Remote albums
- Backup selection page uses the local album details from the DB instead of fetching them during runtime. This change is made in the hope that it would fix the infinite loading issue in the backup album selection page

shenlong-tanwen committed:
refactor(mobile): split albums into remote / local

shenlong-tanwen committed:
refactor(album): bring back backup_albums

fyfrey commented on mobile/lib/modules/album/services/local_album.service.dart:
```diff
@@ -0,0 +1,377 @@
+
+  /// Syncs all device albums and their assets to the database
+  /// Returns `true` if there were any changes
+  Future<bool> _syncLocalAlbumAssetsToDb(List<AssetPathEntity> onDevice) async {
```
Since you moved this from the sync service, it's no longer concurrency safe (the sync service ran this with its internal lock). If a remote sync runs concurrently, the DB state might be garbage.

shenlong-tanwen replied:
Ah, so the same asset could be updated at the same time from the remote sync and local sync resulting in one of two sync details getting lost? Would it be okay if, for the time being, we make the mutex in the sync service public and use it in the local album service as well?

fyfrey replied:
Yes, that would happen. Possibly for a whole batch of assets.

Making the lock public would work fine. Or moving this function back (but I guess you explicitly want to move all handling of local assets here).

shenlong-tanwen committed:
fix: use mutex for sync local albums

fyfrey commented on mobile/lib/modules/backup/background_service/background.service.dart:
```diff
@@ -364,30 +362,17 @@ class BackgroundService {
-          db.backupAlbums.deleteAllSync(toDelete);
-          db.backupAlbums.putAllSync(toUpsert);
-        });
+        // TODO: update album specific last backup time
```
this TODO needs to be resolved

shenlong-tanwen replied:
Yes, but I'm planning to do it in a successive PR. Adding in the TODO just in case I forget to do it.

fyfrey reviewed (commented):
I can't run this PR currently cause my notebook has not enough RAM, so I did only review the code.  Overall, it looks good to me!

alextran1502 committed:
merge main

alextran1502 commented:
Thank you for the PR. One bug I am running into is that on the Sharing page, when you click on an album, it is not loading any assets in the album, a spinner will appear and keep spinning

shenlong-tanwen committed:
refactor: sync shared albums after device albums from homepage

shenlong-tanwen committed:
fix: shared album route pass isarId

shenlong-tanwen commented:
> Thank you for the PR. One bug I am running into is that on the Sharing page, when you click on an album, it does not load any assets in the album; a spinner will appear and keep spinning.

This is fixed now. Let me go through the changes once and make sure that I did not made a similar mistake someplace else as well

> After logging in and navigating to the Sharing page, it takes noticeably longer than `main` for all the albums to appear.
> 
> Can you reproduce the two behaviors above?
> 

I'm not exactly sure why it is taking longer on this branch since nothing is changed related to how a remote album is synced. However, I now changed the order in which we sync assets from the homepage. It used to be remote -> device -> shared and I've updated it to remote -> shared -> device which should make it better

shenlong-tanwen committed:
refactor(mobile): bring back backup selected local assets to timeline (#7090)

* feat(mobile): select which local assets to display in timeline

* remove album selection chips

* refactor: move backup selection to device asset

---------

Co-authored-by: shenlong-tanwen <139912620+shalong-tanwen@users.noreply.github.com>

shenlong-tanwen committed:
chore: pull main

yousefmansy1 commented:
Is this branch being dropped?

It seems like it would resolve a feature discussion here: https://github.com/immich-app/immich/discussions/7075

Thanks

shenlong-tanwen commented:
> Is this branch being dropped?
> 
> 
> 
> It seems like it would resolve a feature discussion here: https://github.com/immich-app/immich/discussions/7075
> 
> 

It is dropped for now. I'll start working on it fresh in a couple of weeks once a proper design is finalised. It won't be available soon, but it is definitely planned in the major refactor of the backup mechanism

yousefmansy1 commented:
Sounds good, thanks for your work on this so far. 👍
